File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0
        
        def isGood(node, maxValue): 
            if not node:
                return 
            print(f"node{node.val} -> {maxValue}")
            nonlocal res
            if node.val >= maxValue:
                print(f"node:{node.val}")
                res += 1
                maxValue = node.val
            if node.left: 
                isGood(node.left, maxValue)    
            if node.right:
                isGood(node.right, maxValue) 

        isGood(root, root.val)        
        return res 


        

File: Data Structures & Algorithms/generate-parentheses/submission-0.java
class Solution {
    List<String> res = new ArrayList<>(); 
    public List<String> generateParenthesis(int n) {
        
        backtrack(n, n, "");
        return res ;
    }


    void backtrack(int left, int right, String s){
        if(s.isEmpty()){
            s ="(";
            left-- ;
        }
        if(left == 0 && right == 0){
            res.add(s);
            return;
        }
        if(left >0){
            backtrack(left-1, right, s + "(");
        }
        if(right > left){
            backtrack(left, right-1, s+ ")");
        }
    }

}
